persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java

```diff
@@ -543,11 +539,39 @@ public boolean contains(Object key) {
          }
          ps.setFetchSize(tableManager.getFetchSize());
          ResultSet rs = ps.executeQuery();
-         return function.apply(rs);
-      }, kvp -> {
-         JdbcUtil.safeClose(kvp.getValue());
-         connectionFactory.releaseConnection(kvp.getKey());
-      });
+         return function.apply(rs).doOnComplete(() -> JdbcUtil.safeClose(rs));
+      }, FlowableConnection::close);
+   }
+
+   class FlowableConnection {
+      final boolean autoCommit;
+      final ConnectionFactory factory;
+      final Connection connection;
+      final PreparedStatement statement;
+
+      FlowableConnection(ConnectionFactory factory, String sql) throws SQLException {
+         this.factory = factory;
+         this.connection = factory.getConnection();
+         this.autoCommit = connection.getAutoCommit();
+         this.statement = connection.prepareStatement(sql);
+
+         // Some JDBC drivers require auto commit disabled to do paging, but we must ensure autoCommit is true in case
```

Comment:
I think this comment is slightly off. I think what it is is that the managed environment prevents these methods from being invoked by throwing an exception.